lib/temporal/client/implementation.rb

```diff
@@ -289,14 +296,26 @@ def process_workflow_result_from(response)
           raise Temporal::Error, 'Workflow execution terminated'
 
         when :EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT
+          attributes = event.workflow_execution_timed_out_event_attributes
+          follow(attributes&.new_execution_run_id) if follow_runs
+
           # TODO: Use more specific error and decode failure
           raise Temporal::Error, 'Workflow execution timed out'
 
         when :EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW
+          attributes = event.workflow_execution_continued_as_new_event_attributes
+          follow(attributes&.new_execution_run_id) if follow_runs
+
           # TODO: Use more specific error and decode failure
           raise Temporal::Error, 'Workflow execution continued as new'
         end
       end
+
+      def follow(new_run_id)
+        return if !new_run_id || new_run_id.empty?
+
+        throw(:next, new_run_id) # next loop with a new run_id
```

Comment:
Is it normal in Ruby to use throw/catch for logic like this? Throwing something in non-erroring cases just to catch it is usually bad practice in other langs. One might expect a loop and/or an explicit return type instead.

Reply:
I think it's the unfortunate naming collision with other languages that's confusing here.

Ruby uses `raise`/`rescue` for handling exceptions (similar to Python's `raise`/`except` and JS's `throw`/`catch`).

Ruby's `throw`/`catch` on the other hand is a control flow mechanism used in situations like this where you need to escape from a nested code. It actually expects every `throw` to be caught, otherwise it will `raise` an `UncaughtThrowError` suggesting a misuse.